src/commands/utils.rs

```diff
@@ -157,23 +191,43 @@ pub fn select_coins_for_spend(
         feerate: FeeRate::from_sat_per_vb(feerate_vb),
         min_fee,
     };
-    if let Err(e) = selector.run_bnb(
-        LowestFee {
-            target,
-            long_term_feerate,
-            change_policy: &change_policy,
-        },
-        100_000,
-    ) {
+    let lowest_fee = LowestFee {
+        target,
+        long_term_feerate,
+        change_policy: &change_policy,
+    };
+    let must_have_change = base_tx.output.is_empty(); // must have change if it is a self-send
+    let lowest_fee_change_cond = LowestFeeChangeCondition {
+        lowest_fee,
+        must_have_change,
+    };
+    if let Err(e) = selector.run_bnb(lowest_fee_change_cond, 100_000) {
         warn!(
             "Coin selection error: '{}'. Selecting coins by descending value per weight unit...",
             e.to_string()
         );
         selector.sort_candidates_by_descending_value_pwu();
-        // If more coins still need to be selected to meet target, then `change_policy(&selector, target)`
-        // will give `Drain::none()`, i.e. no change, and this will simply select more coins until
-        // they cover the target.
-        selector.select_until_target_met(target, change_policy(&selector, target))?;
+        // Select more coins until target is met and change condition satisfied.
+        loop {
+            let drain = change_policy(&selector, target);
+            if selector.is_target_met(target, drain) && (drain.is_some() | !must_have_change) {
```

Comment:
Did you mean to use a binary or here?
```suggestion
            if selector.is_target_met(target, drain) && (drain.is_some() || !must_have_change) {
```

Reply:
Oops, thanks for spotting. That should be a logical or. I'll change it.